Initial implementation review

kinland opened the pull request:
Initial implementation. See the included README for details.

kinland committed:
Add package.json, basic dependencies, and linting scaffolding

kinland committed:
Add 'cy.tget' command to reduce boilerplate [data-test] selectors

kinland committed:
Add interfaceGenerator and related bin executable

kinland committed:
Generate selector shorthands (matching output from interfaceGenerator) and overwrite cy.within

kinland committed:
Add Readme + Markdown linting. Reorg eslint config file.

kinland committed:
Add CODEOWNERS

kinland commented on README.md:
```diff
@@ -1 +1,183 @@
 # cypress-selector-generator
+
+Tired of the boilerplate needed for writing unambiguous selectors for Cypress UI tests? Use `cypress-selector-generator` to create navigation shorthands!
```
This line & the first line have the wrong package name

ViridianForge replied:
For artistic flow - I think you also need another beat.
```suggestion
Are ambiguous Cypress UI test selectors driving you batty? Does the thought of writing boilerplate needed for writing *un*ambiguous selectors for your tests leave you exhausted?  Then get equipped with `cypress-selector-shorthand` and start creating navigation shorthands with absolute ease!
```

Only thing better would be a GIF montage of SDETs flailing on a keyboard with informercial levels of competence.

kinland replied:
Presumably you're being facetious?

ViridianForge replied:
I mean, only if you think you can't make the GIF montage happen.
Consider it an artistic flow suggestion more than anything else. :smile:

kinland commented on lib/global.d.ts:
```diff
@@ -49,9 +70,76 @@ declare global {
+            state<StateKey extends keyof State>(key: StateKey): State[StateKey];
+        }
+
+        // @ts-expect-error No easy solution for this, but if we ignore the typing error, it works as expected.
```
Figured out a solution using both an internal interface and a templated generic type wrapped around it, and will update the export before I merge this PR. Pseudo: `interface _ChainableLikeInternal<A,B> extends Chainable<A,B>{}` wrapped by `export type ChainableLike<A,B> = _ChainableLikeInner<A,B> & B`, so that Typescript doesn't complain about extending the generic type. The resulting IntelliSense mouseover wasn't indecipherable.

kinland commented on package.json:
```diff
@@ -49,6 +56,7 @@
     "lint-staged": "^13.2.3",
     "madge": "^6.1.0",
+    "shx": "^0.3.4",
     "stop-only": "^3.3.1",
```
`shx` can probably be removed.

Future PR will probably swap out `ts-node` for `node` with a transpile/dist step. Not sure `js` consumers could use this directly otherwise.

ViridianForge replied:
If `shx` can be removed, any reason to not do so as part of the initial PR?

kinland replied:
Just didn't notice it had snuck in

kinland commented on lib/install.ts:
```diff
@@ -14,3 +16,28 @@ function calculateTgetSelector(selector: string): string {
+    return cy.wrap(originalWithinFn(...args), noLog)
+        .then(() => {
+            // eslint-disable-next-line no-underscore-dangle
+            delete aliases?.____generatedNavigationWithin;
```
Suggestions for alternate names? When I add a transpile step and actually publish this, I could use the modern `#` for private without breaking older downstreams... but a Cypress alias like `@#genereatedNavigationWithin` to get rid of the underscores still feels odd.

ViridianForge replied:
What specifically feels odd about that? And is `#` generally accepted as a modern standard enough to justify trying the new thing?

kinland replied:
Mostly that I had to disable a linting rule. 

Probably accepted enough: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_class_fields

ViridianForge replied:
I'd say roll with the `#` standard, given that it looks like all browsers have it implemented now, and my own Googling indicated that enough people are using it to have a couple threads about having eslint work with the feature.

kinland replied:
How do you feel about `.as('cypressSelectorGenerator.#generatedNavigationWithin')` / `cy.get('aliases')['cypressSelectorGenerator.#generatedNavigationWithin']` That feels a little weird, too, but the chance of collision should be zero, and it makes it more explicit that we mean # as private and not any other meaning of the symbol in other contexts

ViridianForge replied:
Roll with it!

ViridianForge commented on README.md:
````diff
@@ -1 +1,183 @@
+    });
+```
+
+The above snippets equivalent to the following raw Cypress:
````
* The above snippets are equivalent to ...

ViridianForge replied:
* The above snippets are equivalent to ...

ViridianForge commented on README.md:
````diff
@@ -1 +1,183 @@
+    });
+```
+
+Or, if you would prefer sticking closer to normal Cypress syntax, there's a new `cy.tget` command. It works the same way as the built-in `cy.get`, but automatically wraps non-css/jQuery identifiers with `[data-test]`.
````
Stylistically - would suggest dropping the Or that starts the sentence.

If your package is providing `cy.tget` - I would suggest making that clear with:
"this library provides the new `cy.tget` command."

ViridianForge commented on README.md:
````diff
@@ -1 +1,183 @@
+npx cypress-selector-shorthand interfaces --appName 'MyApp' --infile './schema.json' --outfile './src/schema.ts`
+```
+
+This will create a bunch of interfaces from your schema file (using the `quicktype` library):
````
Nit: Roughly how many is a bunch?

ViridianForge replied:
Also, presumably the schemas will be automatically placed in the correct location for use by `npx`?

kinland replied:
the `--outfile` is so that you can place it in your desired location; there isn't a default.

ViridianForge commented on README.md:
```diff
@@ -1 +1,183 @@
+## Configuration
+To make use of the navigation shorthands, there's three extra steps you need to do to generate the objects and interfaces. (Note: you can use `cy.tget()` without this.)
+
+First, you need to create a schema JSON file, e.g. the following might be used for the example at the top of this file:
```
Would a typical Cypress dev know where to place the schema file?

kinland replied:
Location can be anywhere, since it is specified by `--infile` and has no default

ViridianForge commented on README.md:
```diff
@@ -1 +1,183 @@
+
+You might notice this generated a `with` field that didn't exist in the schema. More on that under the Usage section.
+
+Finally, you need to generate your navigation object:
```
where should this code be placed in order for users to take advantage of the navigation object?

kinland replied:
location doesn't matter, but you need to import it. To reply to several of your questions at once:
If you run
`npx cypress-selector-shorthand interfaces --appName 'MyApp' --infile './myAppSchema.json' --outfile './src/myAppInterface.ts'`

then your schema file must be in `./myAppSchema.json` relative to your project directory, and you have to import `./src/myAppInterface.ts` when you create your `./src/myApp.ts` file.

When I update the PR, I have revised the README and hopefully that will provide some clarity

ViridianForge reviewed (commented):
Warming up with a few thoughts on the README. Kudos for the thoroughness.

kinland committed:
Update README.md

kinland committed:
Move calculateTgetSelector into lib/util.ts

kinland committed:
Fix bug in lib/navigationGenerator.ts/proxifiedGet
When calling Cypress methods chained off the proxy,
we must bind the correct this, or Cypress will try
to get specWindow/chainerId from proxy.

kinland committed:
Fix Types when used as dependency

kinland committed:
Add test from cypress-example-kitchen-sink

kinland committed:
Revamp the Todo test suite using cypress-selector-shorthand

kinland committed:
Add webpack for bundling

kinland committed:
Replace ___ with # as Cypress alias indicator for "private"

kinland committed:
Bump to v0.0.2

kinland commented:
Main TL;DR of the commits I just pushed:
- actually published [on NPM](https://www.npmjs.com/package/cypress-selector-shorthand) now
- there's test cases now
- bundle with webpack
- some bugfixes

kinland committed:
Fix webpack / import issues downstream
Also bump to 0.0.3

kinland committed:
Remove type:module from package.json & remove @cypress/webpack-preprocessor
In 0.0.3, importing without a custom cypress preprocessor worked fine, but these changes seemed to be necessary to make it play nice with a custom webpack config

Incidentally, we don't need to override Cypress's webpack config anymore

kinland committed:
Quote tget selectors to handle parenthesis, etc.

kinland commented on README.md:
````diff
@@ -1 +1,192 @@
+Once you have your schema, you need to generate interfaces for IntelliSense:
+
+```bash
+npx cypress-selector-shorthand interfaces --appName 'MyApp' --infile './myAppSchema.json' --outfile './src/myAppInterface.ts'`
````
```suggestion
npx cypress-selector-shorthand interfaces --appName 'MyApp' --schema './myAppSchema.json' --outfile './src/myAppInterface.ts'`
```

also rename "input" parameter to "schema" and "out-file" parameter to "outfile"
